flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+
+        List<StreamNode> enterPoints =
+                getEnterPoints(sourceNodes, topologicallySortedChainedStreamNodesMap);
+
+        for (StreamNode streamNode : enterPoints) {
+            traverseFullGraph(
+                    streamNode.getId(),
+                    streamNode.getId(),
+                    transitiveNonChainableOutEdgesMap,
+                    topologicallySortedChainedStreamNodesMap,
+                    visitedStartNodes);
+        }
+        computeForwardGroupAndSetNodeParallelisms(
+                transitiveNonChainableOutEdgesMap, topologicallySortedChainedStreamNodesMap);
+    }
+
+    private List<StreamNode> getEnterPoints(
+            List<StreamNode> sourceNodes, Map<StreamNode, List<StreamNode>> chainedStreamNodesMap) {
+        List<StreamNode> enterPoints = new ArrayList<>();
+        for (StreamNode sourceNode : sourceNodes) {
+            if (isSourceChainable(sourceNode, streamGraph)) {
+                StreamEdge outEdge = sourceNode.getOutEdges().get(0);
+                StreamNode startNode = streamGraph.getStreamNode(outEdge.getTargetId());
+                chainedStreamNodesMap
+                        .computeIfAbsent(startNode, k -> new ArrayList<>())
+                        .add(sourceNode);
+                enterPoints.add(startNode);
+            } else {
+                chainedStreamNodesMap.computeIfAbsent(sourceNode, k -> new ArrayList<>());
+                enterPoints.add(sourceNode);
+            }
+        }
+        return enterPoints;
+    }
+
+    private void traverseFullGraph(
+            Integer startNodeId,
+            Integer currentNodeId,
+            Map<Integer, List<StreamEdge>> transitiveNonChainableOutEdgesMap,
+            Map<StreamNode, List<StreamNode>> topologicallySortedChainedStreamNodesMap,
+            Set<Integer> visitedStartNodes) {
+        if (visitedStartNodes.contains(startNodeId)) {
+            return;
+        }
+        List<StreamEdge> chainableOutputs = new ArrayList<>();
+        List<StreamEdge> nonChainableOutputs = new ArrayList<>();
+        StreamNode currentNode = streamGraph.getStreamNode(currentNodeId);
+        StreamNode startNode = streamGraph.getStreamNode(startNodeId);
+        for (StreamEdge streamEdge : currentNode.getOutEdges()) {
+            if (isChainable(streamEdge, streamGraph)) {
+                chainableOutputs.add(streamEdge);
+            } else {
+                nonChainableOutputs.add(streamEdge);
+                topologicallySortedChainedStreamNodesMap.computeIfAbsent(
+                        streamGraph.getStreamNode(streamEdge.getTargetId()),
+                        k -> new ArrayList<>());
+            }
+        }
+
+        transitiveNonChainableOutEdgesMap
+                .computeIfAbsent(startNodeId, k -> new ArrayList<>())
+                .addAll(nonChainableOutputs);
+
+        topologicallySortedChainedStreamNodesMap.get(startNode).add(currentNode);
```

Comment:
Looks to me it is more clear to move this operation to the beginning of this method, after the visited-check?